### PR TITLE
Move e2e tests to separate workspace

### PR DESCRIPTION
**Pull request by kadey001**
e2e tests rely on the whole project, there is no need for playwright to be part of the client workspace and thus should be moved to its own workspace. This will help to reduce the size and complexity of the client and make e2e tests easier to work with.

### CODE REVIEW

File: .github/workflows/development-workflow.yml [hunk 1]
```diff
@@ -81,20 +81,20 @@ jobs:
               run: yarn workspace @app/server generate
```
Comment (DevelopmentByDavid) [inline .github/workflows/development-workflow.yml:86]: I think now is probably a good opportunity to move e2e tests into their own workflow? I'm not sure. I feel like that makes sense but I'm unsure of the tradeoffs. What is your opinion?

One workflow to test the build and another to test the e2e test.  Also, I think it makes sense to potentially scope the trigger based on paths. 

See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

> Use the paths filter when you want to include file path patterns or when you want to both include and exclude file path patterns. Use the paths-ignore filter when you only want to exclude file path patterns. You cannot use both the paths and paths-ignore filters for the same event in a workflow.

File: app/e2e/package.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "@app/e2e",
+    "version": "0.1.0",
+    "private": true,
+    "scripts": {
+        "test:e2e": "playwright test && yarn allure:generate-report",
+        "test:e2e:ci": "dotenv -v CI=1 -- yarn run test:e2e",
+        "test:e2e:headed": "playwright test --headed && yarn allure:generate-report",
+        "allure:generate-report": "allure generate ./allure-results --clean",
+        "allure:report": "allure open ./allure-report"
```
Comment (DevelopmentByDavid) [inline app/e2e/package.json]: do we need to scope these with `test:`?  I think it already makes sense that it's a test since we have to prefix with `yarn workspace @app/e2e`.

Maybe
* test
* ci:test
* debug (test headed because it's for debugging purposes)
* report (I'm not sure why generate-report and report are separate?)
Is there any reason we wouldn't want a clean report every time?

Also, could you update the documentation scripts.md?  It's pretty easy, it's just a markdown file

https://github.com/c2c-project/prytaneum/blob/dev/app/docs/docs/developing/scripts.md

File: app/e2e/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (DevelopmentByDavid) [inline app/e2e/tsconfig.json:1]: Is this the recommended config from playwright or one autogenerated from tsc --init?

**Review by DevelopmentByDavid — CHANGES_REQUESTED** [review]

File: .github/workflows/development-workflow.yml (hunk 1 above)
Comment (kadey001) [reply .github/workflows/development-workflow.yml:86]: We could move it to another workflow but I don't think there would be any real benefit to doing it that way. And I did try to get some of the jobs to only run based on updates to a given path, but I had issues with it sometimes not working correctly (it would either not fire when that path was updated or it would fire when there wasn't an update to that path). I could have been doing something wrong but it seemed to work half the time so it's hard to know for sure.

File: app/e2e/package.json (hunk 2 above)
Comment (kadey001) [reply app/e2e/package.json]: Yeah that makes sense, I'll update them and add to the documentation.

Comment (kadey001) [reply app/e2e/package.json]: Also for the report thing, one opens the report in the browser `allure:report` and the other is the one that actually generates the report. That's why it's two scripts.

File: app/e2e/tsconfig.json (hunk 3 above)
Comment (kadey001) [reply app/e2e/tsconfig.json:1]: No I had just copied that over from the client and edited it. Should I delete it and recreate with tsc --init?

File: app/e2e/package.json (hunk 2 above)
Comment (DevelopmentByDavid) [reply app/e2e/package.json]: Could we potentially consolidate them both into view-report? Or are there times where we would need to generate them, but not view them?

File: .github/workflows/development-workflow.yml (hunk 1 above)
Comment (DevelopmentByDavid) [reply .github/workflows/development-workflow.yml:86]: Okay, we'll keep that out of scope for this PR then. Lets just focus on separating out the workspace we can follow up with workflows later.

File: app/e2e/tsconfig.json (hunk 3 above)
Comment (DevelopmentByDavid) [reply app/e2e/tsconfig.json:1]: Based on the wording here, it seems that typescript should work with playwright by default without a tsconfig.json

See https://playwright.dev/docs/test-typescript#manually-compile-tests-with-typescript
>Sometimes, Playwright Test will not be able to transform your TypeScript code correctly, for example when you are using experimental or very recent features of TypeScript, usually configured in tsconfig.json. **In this case, you can perform your own TypeScript compilation before sending the tests to Playwright. First add a tsconfig.json file inside the tests directory:**

emphasis mine

So I think it should be removed?

File: app/docs/docs/developing/scripts.md [hunk 4]
```diff
@@ -14,6 +14,8 @@ All scripts at the root directory are prefixed with `g:` for global.
 -   `g:start-client`: Start the client with NODE_ENV=production.
 -   `g:client-relay`: Generate `relay` type defs on the client.
 -   `g:start-db`: Start the database using `docker` and `docker-compose`. See `db/start-db.sh`.
+-   `g:start-test-db`: Starts a test specific database using `docker` and `docker-compose`. See `db/start-test-db.sh`.
+-   `g:stop-db`: Stops the currently running database. See `db/stop-db.sh`.
```
Comment (DevelopmentByDavid) [inline app/docs/docs/developing/scripts.md:18]: Nice catch! I'm surprised I didn't already have that in there.

File: app/docs/docs/developing/scripts.md [hunk 5]
```diff
@@ -60,3 +62,23 @@ yarn workspace @app/server add-feat somNewFeature # will add feature/some-new-fe
 -   `prod`: Only used in prod for running the application
 -   `generate`: Generate prisma types. Only necessary to run when the database has been or is being modified and when running the project for the first time.
 -   `lint`: Lint the server.
+-   `prisma-db-push`: Pushes the current prisma schema to the DB.
+-   `prisma-db-seed`: Seeds the DB using the file `app/server/prisma/seed.ts`.
+-   `test:prisma-db-push`: Pushes the current schema with the test `NODE_ENV`.
+-   `test:unit`: Runs unit tests using jest with the test `NODE_ENV`.
+-   `test:integration`: Runs integration tests using jest with the test `NODE_ENV`
```
Comment (DevelopmentByDavid) [inline app/docs/docs/developing/scripts.md]: Could you slightly modify the wording to `...with NODE_ENV=test`. This is more consistent with wording for scripts further up.

File: app/docs/docs/developing/scripts.md [hunk 6]
```diff
@@ -60,3 +62,23 @@ yarn workspace @app/server add-feat somNewFeature # will add feature/some-new-fe
...
+-   `test:ci`: Runs all server tests after ensuring the test DB is running. Used for CI/CD pipeline.
+
+## E2E
```
Comment (DevelopmentByDavid) [inline app/docs/docs/developing/scripts.md]: Maybe order by alpha and further break them up since we're going to have more workspaces in the future probably.
# root
// insert scripts here
# Workspaces
## client
// insert scripts here
## e2e
// insert scripts here
## server
// insert scripts here

File: app/e2e/package.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "@app/e2e",
+    "version": "0.1.0",
+    "private": true,
+    "scripts": {
+        "test": "yarn g:start-test-db && yarn workspace @app/server prisma-db-seed && playwright test && yarn allure:generate-report",
+        "test:ci": "dotenv -v CI=1 -- yarn run test",
+        "debug": "playwright test --headed && yarn allure:generate-report",
+        "allure:generate-report": "allure generate ./allure-results --clean",
+        "allure:open-report": "allure open ./allure-report"
```
Comment (DevelopmentByDavid) [inline app/e2e/package.json]: Only adding a comment here so you remember to respond to my other one re:the allure scripts.  Feel free to close after.

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx [hunk 8]
```diff
@@ -45,13 +45,6 @@ function AskQuestion({ className, eventId }: AskQuestionProps) {
     const { displaySnack } = useSnack();
 
     function handleSubmit(form: TQuestionFormState) {
-        commit({
-            variables: { input: { ...form, eventId, isFollowUp: false, isQuote: false } },
-            onCompleted(payload) {
-                if (payload.createQuestion.isError) displaySnack('Something went wrong!');
-                else close();
-            },
-        });
```
Comment (DevelopmentByDavid) [inline app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx:54]: Was this change intentional or from a rebase?

File: app/e2e/tsconfig.json (hunk 3 above)
Comment (kadey001) [reply app/e2e/tsconfig.json:1]: I tried doing it that way but I get errors when trying to run the tests after they are compiled.
Screenshot from 2022-06-16 17-45-51

Comment (kadey001) [reply app/e2e/tsconfig.json:1]: I'm sure there is a way to fix this, but I'm not sure how much time I should spend on this when it is working the current way. If you want to work on fixing it together sometime though let me know.

File: app/e2e/package.json (hunk 2 above)
Comment (kadey001) [reply app/e2e/package.json]: I mainly had them separated so we could generate and store them in the github actions, that way you can view the results of tests run on there by downloading the report. I could add another script called view report that just runs both if you think that would be better.

File: app/server/src/core/plugins/mercurius.ts [hunk 9]
```diff
@@ -90,7 +90,7 @@ export function attachMercuriusTo(server: FastifyInstance) {
         context: makeRequestContext,
         subscription: {
             context: makeSubscriptionContext,
-            emitter: process.env.NODE_ENV === 'production' ? redisEmitter : undefined,
+            emitter: redisEmitter,
```
Comment (DevelopmentByDavid) [inline app/server/src/core/plugins/mercurius.ts:93]: Leaving a comment to verify that this will now always run, even in development. Is this intended?

**Review by DevelopmentByDavid — APPROVED** [review]
Slight changes, but overall looks good.

File: app/e2e/tsconfig.json (hunk 3 above)
Comment (DevelopmentByDavid) [reply app/e2e/tsconfig.json:1]: We can work on this in another PR. I'll go over the project in general and do some house keeping.

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx (hunk 8 above)
Comment (kadey001) [reply app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx:54]: This was a bug I noticed when I was testing the redis pub/sub locally to see if the emitter was working properly. It was giving adding duplicate questions each time so I quickly fixed it. Figured it was such a small change I didn't want to bother with a whole PR for it.

File: app/server/src/core/plugins/mercurius.ts (hunk 9 above)
Comment (kadey001) [reply app/server/src/core/plugins/mercurius.ts:93]: It should now work with the emitter in development and testing. I'll make sure the students don't have issues running locally though just in case.
